File: week-04-fastapi/app/routers/posts.py
from fastapi import APIRouter, HTTPException, Query
from bson import ObjectId
from datetime import datetime
from app.database import get_db
from app.schemas import PostCreate, PostUpdate, PostResponse

router = APIRouter(prefix="/posts", tags=["posts"])


@router.get("", response_model=dict)
def list_posts(
    page: int = Query(1, ge=1),
    page_size: int = Query(10, ge=1, le=100),
    author_id: str | None = None,
    category_id: str | None = None,
    tag_id: str | None = None,
    published: bool | None = None,
    search: str | None = None,
):
    db = get_db()
    query = {}
    if author_id:
        query["authorId"] = ObjectId(author_id)
    if category_id:
        query["categoryId"] = ObjectId(category_id)
    if tag_id:
        query["tags"] = ObjectId(tag_id)
    if published is not None:
        query["published"] = published
    if search:
        query["$text"] = {"$search": search}

    total = db.posts.count_documents(query)
    posts = (
        db.posts.find(query)
        .sort("createdAt", -1)
        .skip((page - 1) * page_size)
        .limit(page_size)
    )
    items = []
    for p in posts:
        author = db.users.find_one({"_id": p["authorId"]})
        category = db.categories.find_one({"_id": p["categoryId"]})
        p["author_name"] = author["username"] if author else None
        p["category_name"] = category["name"] if category else None
        items.append(PostResponse(**p))

    return {"items": items, "total": total, "page": page, "page_size": page_size}


@router.get("/{post_id}", response_model=PostResponse)
def get_post(post_id: str):
    if not ObjectId.is_valid(post_id):
        raise HTTPException(400, "Invalid post ID")
    db = get_db()
    post = db.posts.find_one({"_id": ObjectId(post_id)})
    if not post:
        raise HTTPException(404, "Post not found")
    return PostResponse(**post)


@router.post("", response_model=PostResponse, status_code=201)
def create_post(body: PostCreate):
    db = get_db()
    if not db.users.find_one({"_id": ObjectId(body.authorId)}):
        raise HTTPException(400, "Author not found")
    if not db.categories.find_one({"_id": ObjectId(body.categoryId)}):
        raise HTTPException(400, "Category not found")

    doc = body.model_dump()
    doc["authorId"] = ObjectId(body.authorId)
    doc["categoryId"] = ObjectId(body.categoryId)
    doc["tags"] = [ObjectId(t) for t in body.tags]
    doc["commentCount"] = 0
    doc["createdAt"] = datetime.now()
    doc["updatedAt"] = datetime.now()

    result = db.posts.insert_one(doc)
    post = db.posts.find_one({"_id": result.inserted_id})
    return PostResponse(**post)


@router.put("/{post_id}", response_model=PostResponse)
def update_post(post_id: str, body: PostUpdate):
    if not ObjectId.is_valid(post_id):
        raise HTTPException(400, "Invalid post ID")
    db = get_db()

    update_data = {k: v for k, v in body.model_dump().items() if v is not None}
    if not update_data:
        raise HTTPException(400, "No fields to update")

    if "categoryId" in update_data:
        update_data["categoryId"] = ObjectId(update_data["categoryId"])
    if "tags" in update_data:
        update_data["tags"] = [ObjectId(t) for t in update_data["tags"]]
    update_data["updatedAt"] = datetime.now()

    result = db.posts.update_one(
        {"_id": ObjectId(post_id)},
        {"$set": update_data}
    )
    if result.matched_count == 0:
        raise HTTPException(404, "Post not found")
    post = db.posts.find_one({"_id": ObjectId(post_id)})
    return PostResponse(**post)


@router.delete("/{post_id}", status_code=204)
def delete_post(post_id: str):
    if not ObjectId.is_valid(post_id):
        raise HTTPException(400, "Invalid post ID")
    db = get_db()
    result = db.posts.delete_one({"_id": ObjectId(post_id)})
    if result.deleted_count == 0:
        raise HTTPException(404, "Post not found")

File: week-05-express/src/routes/posts.js
import { Router } from "express";
import { ObjectId } from "mongodb";
import { getDb } from "../db.js";

const router = Router();

router.get("/", async (req, res) => {
  const db = getDb();
  const { page = 1, page_size = 10, author_id, category_id, tag_id, published, search } = req.query;
  const pageNum = Math.max(1, parseInt(page));
  const pageSize = Math.min(100, Math.max(1, parseInt(page_size)));

  const query = {};
  if (author_id) query.authorId = new ObjectId(author_id);
  if (category_id) query.categoryId = new ObjectId(category_id);
  if (tag_id) query.tags = new ObjectId(tag_id);
  if (published !== undefined) query.published = published === "true";
  if (search) query.$text = { $search: search };

  const total = await db.collection("posts").countDocuments(query);
  const posts = await db.collection("posts").find(query)
    .sort({ createdAt: -1 })
    .skip((pageNum - 1) * pageSize)
    .limit(pageSize)
    .toArray();

  const items = posts.map(p => ({
    id: p._id.toString(),
    title: p.title,
    content: p.content,
    excerpt: p.excerpt,
    authorId: p.authorId.toString(),
    categoryId: p.categoryId.toString(),
    tags: p.tags.map(t => t.toString()),
    commentCount: p.commentCount,
    published: p.published,
    createdAt: p.createdAt,
    updatedAt: p.updatedAt,
  }));

  res.json({ items, total, page: pageNum, page_size: pageSize });
});

router.get("/:id", async (req, res) => {
  if (!ObjectId.isValid(req.params.id)) return res.status(400).json({ detail: "Invalid post ID" });
  const db = getDb();
  const post = await db.collection("posts").findOne({ _id: new ObjectId(req.params.id) });
  if (!post) return res.status(404).json({ detail: "Post not found" });
  res.json({
    id: post._id.toString(),
    title: post.title,
    content: post.content,
    excerpt: post.excerpt,
    authorId: post.authorId.toString(),
    categoryId: post.categoryId.toString(),
    tags: post.tags.map(t => t.toString()),
    commentCount: post.commentCount,
    published: post.published,
    createdAt: post.createdAt,
    updatedAt: post.updatedAt,
  });
});

router.post("/", async (req, res) => {
  const db = getDb();
  const body = req.body;

  if (!await db.collection("users").findOne({ _id: new ObjectId(body.authorId) })) {
    return res.status(400).json({ detail: "Author not found" });
  }
  if (!await db.collection("categories").findOne({ _id: new ObjectId(body.categoryId) })) {
    return res.status(400).json({ detail: "Category not found" });
  }

  const doc = {
    title: body.title,
    content: body.content,
    excerpt: body.excerpt || "",
    authorId: new ObjectId(body.authorId),
    categoryId: new ObjectId(body.categoryId),
    tags: (body.tags || []).map(t => new ObjectId(t)),
    commentCount: 0,
    published: body.published !== undefined ? body.published : true,
    createdAt: new Date(),
    updatedAt: new Date(),
  };

  const result = await db.collection("posts").insertOne(doc);
  const post = await db.collection("posts").findOne({ _id: result.insertedId });
  res.status(201).json({
    id: post._id.toString(),
    title: post.title,
    content: post.content,
    excerpt: post.excerpt,
    authorId: post.authorId.toString(),
    categoryId: post.categoryId.toString(),
    tags: post.tags.map(t => t.toString()),
    commentCount: post.commentCount,
    published: post.published,
    createdAt: post.createdAt,
    updatedAt: post.updatedAt,
  });
});

router.put("/:id", async (req, res) => {
  if (!ObjectId.isValid(req.params.id)) return res.status(400).json({ detail: "Invalid post ID" });
  const db = getDb();
  const body = req.body;

  const update = {};
  if (body.title !== undefined) update.title = body.title;
  if (body.content !== undefined) update.content = body.content;
  if (body.excerpt !== undefined) update.excerpt = body.excerpt;
  if (body.categoryId !== undefined) update.categoryId = new ObjectId(body.categoryId);
  if (body.tags !== undefined) update.tags = body.tags.map(t => new ObjectId(t));
  if (body.published !== undefined) update.published = body.published;
  if (Object.keys(update).length === 0) return res.status(400).json({ detail: "No fields to update" });
  update.updatedAt = new Date();

  const result = await db.collection("posts").updateOne(
    { _id: new ObjectId(req.params.id) },
    { $set: update }
  );
  if (result.matchedCount === 0) return res.status(404).json({ detail: "Post not found" });
  const post = await db.collection("posts").findOne({ _id: new ObjectId(req.params.id) });
  res.json({
    id: post._id.toString(),
    title: post.title,
    content: post.content,
    excerpt: post.excerpt,
    authorId: post.authorId.toString(),
    categoryId: post.categoryId.toString(),
    tags: post.tags.map(t => t.toString()),
    commentCount: post.commentCount,
    published: post.published,
    createdAt: post.createdAt,
    updatedAt: post.updatedAt,
  });
});

router.delete("/:id", async (req, res) => {
  if (!ObjectId.isValid(req.params.id)) return res.status(400).json({ detail: "Invalid post ID" });
  const db = getDb();
  const result = await db.collection("posts").deleteOne({ _id: new ObjectId(req.params.id) });
  if (result.deletedCount === 0) return res.status(404).json({ detail: "Post not found" });
  res.status(204).send();
});

export default router;
